### PR TITLE
Add `directory-` and `script-` adapters

### LINKED ISSUE #4: Hierarchical configuration (a la `config`)
It'd be awesome if we could point `zod-config` at a directory of, say, `.json` files and have it apply configuration from the files following a set load-order, similar to the [`config` package](https://npmjs.com/package/config).

Here's the load-order that package uses: [File Load Order](https://github.com/node-config/node-config/wiki/Configuration-Files#file-load-order)

### LINKED ISSUE #5: JavaScript / TypeScript file adapter
E.g:

```ts
import { z } from 'zod';
import { loadConfig } from 'zod-config';
import { scriptAdapter } from 'zod-config/script-adapter';

const schemaConfig = z.object({
  port: z.string().regex(/^\d+$/),
  host: z.string(),
});

const filePath = path.join(__dirname, 'config.ts');

const config = await loadConfig({
  schema: schemaConfig,
  adapters: [
    scriptAdapter({ path: filePath }),
  ],
});
```

I believe using dynamic imports (i.e. `await import(fileName)`) would yield the desired results:
- In a Javascript runtime, `await import('whatever.ts')` will fail unless the consumer has added a TypeScript loader to the environment
- In a TypeScript runtime, `await import('whatever.ts')` will succeed
- `await import('whatever.js')`, `await import('whatever.mjs')`, and `await import('whatever.cjs')` should always succeed

### PR DESCRIPTION
**Pull request by Lordfirespeed**
- Resolves #4 
- Resolves #5

Motivation / usage example: [Lordfirespeed/test-zod-config-directory](https://github.com/Lordfirespeed/test-zod-config-directory)

### CODE REVIEW

**Comment by Lordfirespeed** [comment]
Oh, yeah - this PR is currently missing tests. I am more than happy to contribute tests, but I'll wait to see whether you approve of the changes in this form, first.

**Comment by alexmarqs** [comment]
Hi @Lordfirespeed awesome, thanks for the contributions and new ideas. I will try to check your changes later (probably later today or during the weekend) 👍! But for sure tests are needed (after checking your code i could also collaborate on that if you need support), I want to make sure we keep the lib with good coverage and quality :).

File: src/lib/adapters/directory-adapter/resolution-and-load-order.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  path,
+  allowedFilenames,
+  extensionToAdapterFactoryMap
+}: ResolveConfigFilesInDirectoryProps): Promise<ConfigResolutionResult[]> {
+  const baseNames = await readdir(path);
+  const results: ConfigResolutionResult[] = [];
+
+  for (const baseName of baseNames) {
+    const dotIndex = baseName.indexOf(".");
+    const fileName = baseName.substring(0, dotIndex);
```
[the remark this answers is not in the document]
Comment (alexmarqs) [reply src/lib/adapters/directory-adapter/resolution-and-load-order.ts:27]: yep let's change it for something like :
```
  const dotIndex = baseName.lastIndexOf(".");
  if (dotIndex === -1) {
    throw new Error("Invalid file name: no extension")
  } 
  const fileName = baseName.substring(0, dotIndex);
  const extension = baseName.substring(dotIndex);
  (...)
```

[the remark this answers is not in the document]
Comment (Lordfirespeed) [reply src/lib/adapters/directory-adapter/resolution-and-load-order.ts:27]: I was actually thinking of something more like 
```ts
for (fileExtension of extensionToAdapterFactoryMap.keys()) {
  if (!baseName.endsWith(fileExtension) continue;

const fileName = baseName.substring(0, baseName.length - fileExtension.length)
// do stuff with fileName and fileExtension 
}
```
Which handles the case a file extension contains more than one `.` - what do you think?

[the remark this answers is not in the document]
Comment (alexmarqs) [reply src/lib/adapters/directory-adapter/resolution-and-load-order.ts:34]: yep, it works as well! feel free to add it. I have committed some basic script adapter unit tests. Btw, the tests file is quite big already as we are adding more built in adapters but ignore it for now, I will split the tests files later.

[the remark this answers is not in the document]
Comment (alexmarqs) [reply src/lib/adapters/directory-adapter/resolution-and-load-order.ts]: @Lordfirespeed i have committed basic tests for the dir adapter + small refactorings 👍

**Comment by Lordfirespeed** [comment]
Seems we ended up both doing a similar thing :joy: - feel free to mess around/rearrange/remove what I've added

**Comment by Lordfirespeed** [comment]
> thanks! If no further requirements / changes from your side, I can approve the PR 👍

Everything LGTM! <3

**Pull request merged**